src/clj/coffi/mem.clj

```diff
@@ -1350,3 +1766,523 @@
+  (let [aligned-fields
+        (loop [offset 0
+               aligned-fields []
+               fields (nth struct-spec 1)]
+          (if (seq fields)
+            (let [[[_ type :as field] & fields] fields
+                  size (size-of type)
+                  align (align-of type)
+                  r (rem offset align)]
+              (recur (cond-> (+ offset size)
+                       (pos? r) (+ (- align r)))
+                     (cond-> aligned-fields
+                       (pos? r) (conj [:coffi.layout/padding [:coffi.mem/padding (- align r)]])
+                       :always (conj field))
+                     fields))
+            (let [strongest-alignment (reduce max (map (comp align-of second) (nth struct-spec 1)))
+                  r (rem offset strongest-alignment)]
+              (cond-> aligned-fields
+                (pos? r) (conj [:coffi.layout/padding [:coffi.mem/padding (- strongest-alignment r)]])))))]
+    (assoc struct-spec 1 aligned-fields)))
+
+(defn- coffitype->typename [in]
+  (let [[arr _type n & {:keys [raw?] :as opts}] (if (vector? in) in [:- in])
+        arr? (= arr ::array)
+        array-types  {::byte   'bytes
+                      ::short  'shorts
+                      ::int    'ints
+                      ::long   'longs
+                      ::char   'chars
+                      ::float  'floats
+                      ::double 'doubles}
+        single-types {::byte     'byte
+                      ::short    'short
+                      ::int      'int
+                      ::long     'long
+                      ::char     'char
+                      ::float    'float
+                      ::double   'double
+                      ::c-string 'String}]
+    (cond (and arr? raw?) (get array-types _type 'objects)
+          (and arr?)      `clojure.lang.IPersistentVector
+          :default        (get single-types _type (keyword (str *ns*) (str _type))))))
+
+(defn- coffitype->array-fn [_type]
+  (get
+   {:coffi.mem/byte   `byte-array
+    :coffi.mem/short  `short-array
+    :coffi.mem/int    `int-array
+    :coffi.mem/long   `long-array
+    :coffi.mem/char   `char-array
+    :coffi.mem/float  `float-array
+    :coffi.mem/double `double-array}
+   _type
+   `object-array))
+
+(defn- coffitype->array-write-fn [_type]
+  ({:coffi.mem/byte   `write-bytes
+    :coffi.mem/short  `write-shorts
+    :coffi.mem/int    `write-ints
+    :coffi.mem/long   `write-longs
+    :coffi.mem/char   `write-chars
+    :coffi.mem/float  `write-floats
+    :coffi.mem/double `write-doubles} _type))
+
+(defn- coffitype->array-read-fn [_type]
+  ({:coffi.mem/byte   `read-bytes
+    :coffi.mem/short  `read-shorts
+    :coffi.mem/int    `read-ints
+    :coffi.mem/long   `read-longs
+    :coffi.mem/char   `read-chars
+    :coffi.mem/float  `read-floats
+    :coffi.mem/double `read-doubles} _type))
+
+(defmulti  generate-deserialize (fn [& xs] (if (vector? (first xs)) (first (first xs)) (first xs))))
```

Comment:
`ffirst` is a cool function :)

Reply:
fair point. fixed in [10b8baa](https://github.com/IGJoshua/coffi/pull/14/commits/10b8baafa2f09909aaf98f967ca1a5b06f11b97c)